File: sdk/haskell/src/The0/State.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module      : The0.State
-- Description : Persistent state management for the0 trading bots
-- License     : Apache-2.0
--
-- This module provides persistent state management for bots across executions.
-- State is automatically synced to MinIO storage between bot runs.
--
-- = Example
--
-- @
-- import The0.State as State
-- import Data.Aeson (object, (.=))
--
-- main :: IO ()
-- main = do
--     -- Store state
--     State.set "portfolio" (object ["AAPL" .= (100 :: Int), "GOOGL" .= (50 :: Int)])
--
--     -- Retrieve state
--     portfolio <- State.get "portfolio"
--
--     -- List all keys
--     keys <- State.list
--
--     -- Delete a key
--     deleted <- State.delete "portfolio"
--
--     -- Clear all state
--     State.clear
-- @
module The0.State
    ( get
    , set
    , The0.State.delete
    , list
    , clear
    , exists
    ) where

import Control.Exception (catch, SomeException)
import Data.Aeson (FromJSON, ToJSON, decode, encode)
import qualified Data.ByteString.Lazy as BL
import Data.Maybe (fromMaybe)
import System.Directory (createDirectoryIfMissing, doesFileExist, listDirectory, removeFile)
import System.Environment (lookupEnv)
import System.FilePath ((</>), takeExtension, dropExtension)

-- | Get the path to the state directory.
getStateDir :: IO FilePath
getStateDir = fromMaybe "/state/.the0-state" <$> lookupEnv "STATE_DIR"

-- | Get the file path for a state key.
getKeyPath :: String -> IO FilePath
getKeyPath key = do
    stateDir <- getStateDir
    return $ stateDir </> (key ++ ".json")

-- | Validate that a key is safe to use as a filename.
validateKey :: String -> IO ()
validateKey key
    | null key = ioError $ userError "State key cannot be empty"
    | '/' `elem` key || '\\' `elem` key || ".." `isInfixOf` key =
        ioError $ userError "State key cannot contain path separators or '..'"
    | otherwise = return ()
  where
    isInfixOf needle haystack = needle `elem` [take (length needle) (drop i haystack) | i <- [0..length haystack - length needle]]

-- | Get a value from persistent state.
--
-- Returns 'Nothing' if the key doesn't exist or deserialization fails.
--
-- @
-- portfolio <- State.get "portfolio" :: IO (Maybe Portfolio)
-- @
get :: FromJSON a => String -> IO (Maybe a)
get key = do
    validateKey key
    filepath <- getKeyPath key
    exists' <- doesFileExist filepath
    if not exists'
        then return Nothing
        else do
            content <- BL.readFile filepath `catch` handleError
            return $ decode content
  where
    handleError :: SomeException -> IO BL.ByteString
    handleError _ = return ""

-- | Set a value in persistent state.
--
-- The value must be serializable to JSON.
--
-- @
-- State.set "portfolio" (object ["AAPL" .= (100 :: Int)])
-- State.set "trade_count" (42 :: Int)
-- @
set :: ToJSON a => String -> a -> IO ()
set key value = do
    validateKey key
    stateDir <- getStateDir
    createDirectoryIfMissing True stateDir
    filepath <- getKeyPath key
    BL.writeFile filepath (encode value)

-- | Delete a key from persistent state.
--
-- Returns 'True' if the key existed and was deleted, 'False' otherwise.
--
-- @
-- deleted <- State.delete "old_data"
-- when deleted $ putStrLn "Cleaned up old data"
-- @
delete :: String -> IO Bool
delete key = do
    validateKey key
    filepath <- getKeyPath key
    exists' <- doesFileExist filepath
    if exists'
        then do
            removeFile filepath `catch` handleError
            return True
        else return False
  where
    handleError :: SomeException -> IO ()
    handleError _ = return ()

-- | List all keys in persistent state.
--
-- @
-- keys <- State.list
-- putStrLn $ "State contains " ++ show (length keys) ++ " keys"
-- @
list :: IO [String]
list = do
    stateDir <- getStateDir
    files <- listDirectory stateDir `catch` handleError
    return [dropExtension f | f <- files, takeExtension f == ".json"]
  where
    handleError :: SomeException -> IO [FilePath]
    handleError _ = return []

-- | Clear all state.
--
-- Removes all stored state keys.
--
-- @
-- State.clear
-- putStrLn "All state cleared"
-- @
clear :: IO ()
clear = do
    stateDir <- getStateDir
    files <- listDirectory stateDir `catch` handleListError
    mapM_ removeJsonFile files
  where
    handleListError :: SomeException -> IO [FilePath]
    handleListError _ = return []

    removeJsonFile :: FilePath -> IO ()
    removeJsonFile f = do
        stateDir <- getStateDir
        let filepath = stateDir </> f
        if takeExtension f == ".json"
            then removeFile filepath `catch` handleRemoveError
            else return ()

    handleRemoveError :: SomeException -> IO ()
    handleRemoveError _ = return ()

-- | Check if a key exists in state.
--
-- @
-- hasPortfolio <- State.exists "portfolio"
-- when hasPortfolio $ do
--     portfolio <- State.get "portfolio"
--     -- ...
-- @
exists :: String -> IO Bool
exists key = do
    validateKey key
    filepath <- getKeyPath key
    doesFileExist filepath

File: sdk/python/the0/state.py
"""
the0 State Module
=================

Provides persistent state management for bots across executions.
State is automatically synced to MinIO storage between bot runs.

Example:
    from the0 import state

    # Store state
    state.set("portfolio", {"AAPL": 100, "GOOGL": 50})

    # Retrieve state
    portfolio = state.get("portfolio", {})

    # List all keys
    keys = state.list()

    # Delete a key
    state.delete("portfolio")

    # Clear all state
    state.clear()
"""

import os
import json
from typing import Any, List, Optional


def _get_state_dir() -> str:
    """Get the path to the state directory."""
    state_dir = os.environ.get("STATE_DIR")
    if state_dir:
        return state_dir
    # Fallback to default location
    return "/state/.the0-state"


def _get_key_path(key: str) -> str:
    """Get the file path for a state key."""
    state_dir = _get_state_dir()
    return os.path.join(state_dir, f"{key}.json")


def _validate_key(key: str) -> None:
    """Validate that a key is safe to use as a filename."""
    if not key:
        raise ValueError("State key cannot be empty")
    # Prevent directory traversal
    if "/" in key or "\\" in key or ".." in key:
        raise ValueError("State key cannot contain path separators or '..'")


def get(key: str, default: Any = None) -> Any:
    """
    Get a value from persistent state.

    Args:
        key: The state key (alphanumeric, hyphens, underscores)
        default: Default value if key doesn't exist

    Returns:
        The stored value, or default if not found

    Example:
        portfolio = state.get("portfolio", {})
        trade_count = state.get("trade_count", 0)
    """
    _validate_key(key)
    filepath = _get_key_path(key)
    try:
        with open(filepath, "r") as f:
            return json.load(f)
    except FileNotFoundError:
        return default
    except json.JSONDecodeError:
        return default


def set(key: str, value: Any) -> None:
    """
    Set a value in persistent state.

    The value must be JSON serializable.

    Args:
        key: The state key (alphanumeric, hyphens, underscores)
        value: The value to store (must be JSON serializable)

    Example:
        state.set("portfolio", {"AAPL": 100, "GOOGL": 50})
        state.set("trade_count", 42)
        state.set("last_prices", [45000.5, 45100.0, 45050.25])
    """
    _validate_key(key)
    state_dir = _get_state_dir()
    os.makedirs(state_dir, exist_ok=True)
    filepath = _get_key_path(key)
    with open(filepath, "w") as f:
        json.dump(value, f)


def delete(key: str) -> bool:
    """
    Delete a key from persistent state.

    Args:
        key: The state key to delete

    Returns:
        True if the key existed and was deleted, False otherwise

    Example:
        if state.delete("old_data"):
            print("Cleaned up old data")
    """
    _validate_key(key)
    filepath = _get_key_path(key)
    try:
        os.remove(filepath)
        return True
    except FileNotFoundError:
        return False


def list() -> List[str]:
    """
    List all keys in persistent state.

    Returns:
        List of state keys

    Example:
        keys = state.list()
        print(f"State contains {len(keys)} keys: {keys}")
    """
    state_dir = _get_state_dir()
    try:
        files = os.listdir(state_dir)
        return [f[:-5] for f in files if f.endswith(".json")]
    except FileNotFoundError:
        return []


def clear() -> None:
    """
    Clear all state.

    Removes all stored state keys.

    Example:
        state.clear()
        print("All state cleared")
    """
    state_dir = _get_state_dir()
    try:
        for filename in os.listdir(state_dir):
            if filename.endswith(".json"):
                os.remove(os.path.join(state_dir, filename))
    except FileNotFoundError:
        pass


def exists(key: str) -> bool:
    """
    Check if a key exists in state.

    Args:
        key: The state key to check

    Returns:
        True if the key exists, False otherwise

    Example:
        if state.exists("portfolio"):
            portfolio = state.get("portfolio")
    """
    _validate_key(key)
    filepath = _get_key_path(key)
    return os.path.exists(filepath)
